### ISSUE TITLE
Invalid parsing of interpolated strings with curly braces followed after interpolation with format specifier

### ISSUE THREAD

**Issue by bernd5** [dotnet/roslyn#57750]
**Version Used**: commit 973b3aa

**Steps to Reproduce**:
Run the following code with net 6.0:

```csharp
using System;

Console.WriteLine($"{{{12:X}}}");
```
See [here](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWAChis8BOACgBIAiAb0bxxAA0BfT2gSgG4g==)

**Expected Behavior**:
It should print "{C}" like it did in net5.0

**Actual Behavior**:
It prints: "{X}}"

That there is parsing error is very good visible in Visual studios highlighting:
![grafik](https://user-images.githubusercontent.com/26689688/141613850-c04dff63-eb5a-450e-9c3c-1f5ba9bf898c.png)

See csharplang discussion: [here](https://github.com/dotnet/csharplang/discussions/4361)

------

### Details

With net5.0 it was lowered to:

```csharp
Console.WriteLine(string.Format("{{{0:X}}}", 12)); //prints "{C}"
```

but with net6.0 it is lowered to

```csharp
DefaultInterpolatedStringHandler defaultInterpolatedStringHandler = new DefaultInterpolatedStringHandler(1, 1);
defaultInterpolatedStringHandler.AppendLiteral("{");
defaultInterpolatedStringHandler.AppendFormatted(12, "X}}");
Console.WriteLine(defaultInterpolatedStringHandler.ToStringAndClear()); //prints "{X}}"
```

instead of:

```csharp
DefaultInterpolatedStringHandler defaultInterpolatedStringHandler = new DefaultInterpolatedStringHandler(2, 1);
defaultInterpolatedStringHandler.AppendLiteral("{");
defaultInterpolatedStringHandler.AppendFormatted(12, "X");
defaultInterpolatedStringHandler.AppendLiteral("}");
Console.WriteLine(defaultInterpolatedStringHandler.ToStringAndClear()); //prints "{C}"
```

**Comment by CyrusNajmabadi** [comment]
Interesting case. The interpretation on the language side has remained the same. But the lowering impacts how the runtime interprets this. @333fred thoughts?

**Comment by bernd5** [comment]
The [grammar ](https://github.com/dotnet/csharplang/blob/main/spec/lexical-structure.md#interpolated-string-literals) looks like this:

```antlr
interpolation_format
    : ':' interpolation_format_character+
    ;
    
interpolation_format_character
    : '<Any character except \" (U+0022), : (U+003A), { (U+007B) and } (U+007D)>'
    ;
```

Which explicitly defines a `}` as not allowed in the format specifier. See the attached PR.

**Comment by 333fred** [comment]
Is this not the same issue you raised in https://github.com/dotnet/csharplang/discussions/4361?

**Comment by CyrusNajmabadi** [comment]
Right. The question is if the impl matches the intent or not. It's clear that the impl thinks the format specifier should be allowed to have these *escaped* characters in them. And the new handler code respects that. 

However, if it's the case that the underlying runtime code does not see it this way, then an argument could be made to change things. 

Where it gets complex is that this can affect other cases.  There may be code out there that depends on the old interpretation.

**Comment by bernd5** [comment]
Yes, and Cyrus wrote:
> this has already been fixed. In the future, please file issues at roslyn, not csharplang.

But it actually hasn't fixed. I don't see any case where this might be useful.
And the spec explicitly prohibits this - so we should follow the spec.

The reason for all this has more a historic background. In the past `String.Format` did not support curly braces after a "hole".

**Comment by CyrusNajmabadi** [comment]
> And the spec explicitly prohibits this - so we should follow the spec.

No. We should do the right thing for customers, including understanding the problems if we change this. 

If the spec and impl disagree, we may decide to change the spec.  

That said, given the runtime changed here, I'm amenable to a discussion about syncing roslyn to match it ( since the larger breaking change at runtime decision was already approved). However, we have to make that decision first and make sure all stakeholders think it's the right thing to do before making a change

**Comment by bernd5** [comment]
In the very unlikly event there is really a need for closing curly braces in the format specifier we may could allow: "\u007D" there - because this is today not supported in interpolated strings - if you write for example: 

```csharp
Console.WriteLine($"\u007D\u007D");
```
or 

```csharp
Console.WriteLine($"{12:\u007D}");
```

or 

```csharp
Console.WriteLine($"{12:\u007D\u007D}");
```

the compiler generates:

> error CS8087: A '}' character may only be escaped by doubling '}}' in an interpolated string.

**Comment by CyrusNajmabadi** [comment]
Imo, due to the runtime change, I think we have the space to make an equivalent change here as well. Where I'm a little less clear is if we should treat this as a lang change that should go through ldm, or just a compiler change. 

I think there's a strong argument to be made that is only sensical for the lang to match the runtime interpretation. That's why we had the original logic. If the runtime has changed, we should do. 

I'll work with @333fred and @jaredpar to determine the best way forward to make decisions here. Thanks very much!

**Comment by AdamSpeight2008** [comment]
Just to note that visual basic displays `Console.WriteLine($"{{{12:X}}}")` as `{C}`, which I think is the intended behaviour .

**Comment by jaredpar** [comment]
> Where I'm a little less clear is if we should treat this as a lang change that should go through ldm, or just a compiler change.

It's a blurry line. My tendency is to think of it from the vantage of "it's a regression, is it one that we need to fix?" This seems like one we should fix. Generally for cases like this I prefer to move forward with the fix + send an alert / email to LDM in case someone wants to push back on the regression angle. Basically give LDM the opportunity to say "no" vs. waiting for them to say "yes". 

One item I think is missing from this discussion though is what is the option for maintaining the old behavior on the new implementation? Basically how do you get the desired behavior here? And probably most importantly how will that solution work in older compilers?

**Comment by CyrusNajmabadi** [comment]
Feels like there's no good answer here (and even a flag to somewhatr control behavior sounds awful).  Maybe just throw our hands up in the air, 'mea culpa' this and say: sorry if this change causes pain.  here's what happened.  here's waht it means now and for the future.  hopefully you're not impacted.  if you are, sorry.

:(

**Comment by svick** [comment]
@jaredpar 

> One item I think is missing from this discussion though is what is the option for maintaining the old behavior on the new implementation? Basically how do you get the desired behavior here?

Is there any evidence that anyone would actually want this? Did anyone report an issue when the behavior of `string.Format` changed in .Net Core 3.1?

Also, it seems the behavior in .Net 6.0 is not even the same as it was in .Net Framework/.Net Core 2.1. The one-liner in the first post produces the following output:

```
> dotnet run -f net48
{X}
> dotnet run -f netcoreapp2.1
{X}
> dotnet run -f netcoreapp3.1
{C}
> dotnet run -f net5.0       
{C}
> dotnet run -f net6.0       
{X}}
```

Notice how on .Net 6, it prints two closing braces, while older frameworks output only one.

> And probably most importantly how will that solution work in older compilers?

I don't see the problem. Older compilers delegate to `string.Format`, so their behavior depends on the runtime. Once this bug is fixed, new compilers will have the same behavior as old compilers on all runtimes (just optimized using interpolated string handlers where available).

The only problematic case I can think of is if someone uses interpolated string handlers on an older framework. But I don't think that's a case that's worth worrying about.

**Comment by jaredpar** [comment]
> Is there any evidence that anyone would actually want this? 

It's less about want but more about making sure the full set of options is understood. A common question that comes up in discussions like this is "how do you get the desired behavior with the current implementation?".  The cost of the change is one of the variables we consider along side change customers actually hit this, cost of changing the language, etc ... 

In this particular case I fully expect we'll just fix the implementation. At the same time though, it helps to have all the information there in case it comes up. That is particularly important if we decide to push for servicing (patch 17.0 vs. waiting for 17.1 to get released). Servicing has a higher bar and one of the questions that will **absolutely** get asked is 

> How can customers work around this today?

Basically what options are available to customers in 17.0 that could be employed as a stop gap until 17.1? It's considerably less risky to fix a change in a new release vs. servicing an existing one. Hence this is one of questions that comes up

**Comment by bernd5** [comment]
Some workarounds customers can use today are:

```csharp
using System;

Console.WriteLine('{' + $"{12:X}" + '}');
Console.WriteLine($"{"{"}{12:X}{"}"}");
Console.WriteLine($"{{{12:X}{"}"}");
Console.WriteLine(String.Concat('{', String.Format("{0:X}", 12), '}'));

defaultInterpolatedStringHandler = new DefaultInterpolatedStringHandler(0, 3);
defaultInterpolatedStringHandler.AppendLiteral("{");
defaultInterpolatedStringHandler.AppendFormatted(12, "X");
defaultInterpolatedStringHandler.AppendLiteral("}");
Console.WriteLine(defaultInterpolatedStringHandler.ToStringAndClear());
```

**Comment by Timovzl** [comment]
> Is there any evidence that anyone would actually want this? Did anyone report an issue when the behavior of string.Format changed in .Net Core 3.1?

@svick Yes sir: https://github.com/dotnet/runtime/issues/61838.

We use string formats like this to report classes with some of their most identifying properties:

`public override string ToString() => $"{{{this.GetType().Name} Id={this.Id} Name={this.Name}}}"`

The result is a nice and clean `{Person Id=123 Name=Joe}`.

Those curly braces make it very easy to create readable messages mentioning such objects:

`$"{inviter} invited {invitee} to {team}."`

The results in something like `{Person Id=123 Name=Joe} invited {Person Id=234 Name=Newbie} to {Team Name=Example}`.

Even objects mentioning nested objects stay readable this way, since the curly braces clearly bound each object: `{Team Name=Example Leader={Person Id=123 Name=Joe}}`.

However, this approach now creates an undesirable string when the last property uses a format specifier.

I believe the use case is strong. I'm curious about the use case for interpreting `}}` as part of the format specifier.

**Comment by svick** [comment]
@Timovzl I understand that and agree with you. I was asking whether someone was broken when the bug was fixed previously in .Net Core 3.1, to gauge if that is a problem worth mitigating now that the bug needs to be fixed again.
